### ISSUE TITLE
[HOLD for payment 2023-08-28] [$1000] Chat - Email is pasted twice in the composer when copied from user profile

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#25409]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to staging.new.expensify.com
2. Open profile page for any user in chat
3. Click Copy to clipboard icon next to the email
4. Paste the email in the composer

## Expected Result:
Email is pasted once in the composer

## Actual Result:
Email is pasted twice in the composer

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] Windows / Chrome 
- [ ] MacOS / Desktop

**Version Number:** 1.3.55-1

**Reproducible in staging?:** Yes

**Reproducible in production?:** No

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/4ee215d5-74a9-4cd9-8b57-31a86eadbae3

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01f8779e48a7b14b9d</li>
        <li>Upwork Job ID: 1692204576618807296</li>
        <li>Last Price Increase: 2023-08-17</li>
<li>Automatic offers: </li>
<ul>
<li>situchan | Reviewer | 26181155</li>
<li>daordonez11 | Contributor | 26181158</li>
</ul></li>
    </ul>
</details>

**Comment by situchan** [comment]
This was first reported by me - https://expensify.slack.com/archives/C049HHMV9SM/p1692127516386599

**Comment by daordonez11** [comment]
This error was introduce in this PR https://github.com/Expensify/App/pull/23359
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Double paste in composer after copying email prom profile

### What is the root cause of that problem?
Firstly it is important to note that the error only happens when profile is opened (copy email) and then close. because it triggers the handler of [navigation focus](https://github.com/Expensify/App/blob/2e1c4e799977a3c6d276683b8b9959d45255caf7/src/components/Composer/index.js#L358).

Then there are 2 different paste handlers at the same time, one associated to the whole document and one associated to the [input](https://github.com/Expensify/App/blob/2e1c4e799977a3c6d276683b8b9959d45255caf7/src/components/Composer/index.js#L366)

Hence handlePaste is called 2 times  
### What changes do you think we should make in order to solve the problem?
Both handlers should be handled the same way, either both are associated with the document to ensure the handler is removed when navigation blur is called. 

We could also save the textInput handler in a variable if it exists everything should be handled in textInput instead of document and vice-versa.

Also in the return method of the effect we should do: `textInput.current.removeEventListener('paste', handlePaste);` or include the same validation to remove the respective handler.

In pseudocode something like:

'''js
const unsubscribeFocus = navigation.addListener('focus', () => {
            if (textInput.current) {
                textInput.current.addEventListener('paste', handlePaste);
            }
            else {
                document.addEventListener('paste', handlePaste)
            }
        });
        const unsubscribeBlur = navigation.addListener('blur', () => {
            if (textInput.current) {
                textInput.current.removeEventListener('paste', handlePaste);
            }
            else {
                document.removeEventListener('paste', handlePaste)
            }
        });
'''


### What alternative solutions did you explore? (Optional)

Validate if handlePAste is already called in some way and avoid multicall or something like throttle in the same method but too complex for a simple task

cc @mountiny

**Comment by StevenKKC** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Email is pasted twice in the composer when copied from user profile.

### What is the root cause of that problem?
In `Composer` component, we add two `paste` event listener.
https://github.com/Expensify/App/blob/e89ea9d73158ac395a411eca908733a681f598c8/src/components/Composer/index.js#L358
https://github.com/Expensify/App/blob/e89ea9d73158ac395a411eca908733a681f598c8/src/components/Composer/index.js#L366
So the text is pasted twice in the composer.
### What changes do you think we should make in order to solve the problem?
We should remove one `paste` event listener.
Since adding event listener to `textInput.current` is redundant, we can remove [this line](https://github.com/Expensify/App/blob/e89ea9d73158ac395a411eca908733a681f598c8/src/components/Composer/index.js#L366).
```js
    if (textInput.current) {
-       textInput.current.addEventListener('paste', handlePaste);
```
### What alternative solutions did you explore? (Optional)
None.

**Comment by mountiny** [comment]
@daordonez11 Are you bale to create a PR for this with a fix now? Its a deploy blocker so urgency is required. 

@bondydaa would you like to let the contributor fix this?

**Comment by daordonez11** [comment]
I can definitely create a PR and work on it right now @mountiny  @bondydaa

**Comment by mountiny** [comment]
@bondydaa up to you

**Comment by getusha** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Text is pasted twice when text is copied from user profile.
### What is the root cause of that problem?
When navigation is focused it will add a 'paste' event listener, which makes the component have two event listeners at once triggering paste twice.

### What changes do you think we should make in order to solve the problem?

Prior to adding the event listener to 'textInput', initiate a call to `unsubscribeFocus` in order to remove the previous event listener from the document.


### What alternative solutions did you explore? (Optional)

**Comment by bondydaa** [comment]
also FYI discussing this here https://expensify.slack.com/archives/C049HHMV9SM/p1692127516386599

**Comment by situchan** [comment]
@daordonez11 I'd say redundant code already existed after #21583.
This code was just moved while hook refactor.
Can you find out the root cause why this didn't happen before?

**Comment by getusha** [comment]
@situchan the cause is before the event handler is being added on document only for both cases, but now we are adding two events in to separate places 1. on document 2. on textinput which makes it to have two event listeners instead of one. but if it was only on the document or on textinput it won't happen.

we should add the event listener only on `textInput.current` or `document` only depending on the ref. another alternative way is to remove the document listener before adding another one.

**Comment by situchan** [comment]
@daordonez11 that code was just moved from componentDidMount to useEffect with empty array dependency

<img width="1181" alt="Screenshot 2023-08-17 at 10 13 58 PM" src="https://github.com/Expensify/App/assets/108292595/f65f2a1c-e341-42c0-b06c-800fe894985e">

**Comment by Pujan92** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Email pasted twice on paste

### What is the root cause of that problem?
Seems we are having two addEventListener on different elements (on `document` and on  `textInput.current`), due to that it will execute 2 paste handle and causes the issue.

### What changes do you think we should make in order to solve the problem?
Like earlier we can add `addEventListener` on `document` here which makes sure to ignore the second one if the same function we try to set on the same event.
https://github.com/Expensify/App/blob/e89ea9d73158ac395a411eca908733a681f598c8/src/components/Composer/index.js#L366

**Comment by daordonez11** [comment]
@situchan  you are right but the causing line is `textInput.current.addEventListener('paste', handlePaste);` previously it was only used for wheel! and then textInput.current. was used for paste as well. This was introduced during the migration to function component

<img width="932" alt="image" src="https://github.com/Expensify/App/assets/13544910/e4e91e0a-675b-43e5-82ee-351e3a47a541">

Previously paste event was only used in the document context.

**Comment by situchan** [comment]
@Pujan92 thanks for the proposal. How about using `textInput.current.addEventListener` entirely instead of `document.addEventListener`?

**Comment by getusha** [comment]
@situchan i commented the idea earlier https://github.com/Expensify/App/issues/25409#issuecomment-1682575848

**Comment by daordonez11** [comment]
@situchan This is what I mention in the first part of my proposal "Both handlers should be handled the same way, either both are associated with the document to ensure the handler is removed when navigation blur is called or there is a validation to handle each case." . Now seeing the original class component that you shared I can see it is just an error in the migration the paste handler should only be used with document.

**Comment by Pujan92** [comment]
> @Pujan92 thanks for the proposal. How about using `textInput.current.addEventListener` entirely instead of `document.addEventListener`?

I think that will work too, we just need to add them inside the `if (textInput.current)` condition

**Comment by situchan** [comment]
@daordonez11 ah yes you first commented that idea. But I had just got confused with pseudocode.
Let's assign @daordonez11 as they first [proposed](https://github.com/Expensify/App/issues/25409#issuecomment-1682516029) to make listeners consistent

cc: @bondydaa 
🎀 👀 🎀 C+ reviewed

**Comment by Pujan92** [comment]
> "Both handlers should be handled the same way, either both are associated with the document to ensure the handler is removed when navigation blur is called or there is a validation to handle each case." . 

@daordonez11 just to reply, handlers should be handled on the same target that is fine but solution reasoning seems incorrect to me. Basically, we can add multiple events with `addEventListener` but here the thing is we can't set the same function twice for the same target event. In that case the 2nd duplicate one will be ignored.

**Comment by daordonez11** [comment]
Yeah you are right @Pujan92 ! I didn't know about the double reference you mentioned in your proposal :O . I was referring actually to the removeEventListener on the blur which was already pointing to document, hence if the initial one points to the same `document` handler it is removed when the user profile is opened. I was not talking about double `addEventListener` in the same element. Sorry for the lack of details I was trying to do a fast proposal. 😅

**Comment by daordonez11** [comment]
Hey guys so I know this is blocker so here is the PR ready for review https://github.com/Expensify/App/pull/25419 I'm still finishing tests on native but everything looking good in terms of web mweb and desktop.

cc @situchan @mountiny @bondydaa

**Comment by bondydaa** [comment]
great thanks, taking a look!

**Comment by bondydaa** [comment]
merged the PR, asked to get it CP'd internally here https://expensify.slack.com/archives/C07J32337/p1692294385642269

**Comment by b4s36t4** [comment]
<img width="456" alt="Screenshot 2023-08-17 at 11 29 42 PM" src="https://github.com/Expensify/App/assets/59088937/5e86c6eb-e45b-482a-91d5-4fca1249a55d">
 getting this error when tried to paste something without composer focus, is this intended?

**Comment by b4s36t4** [comment]
Just pulled.. checked after pull.

**Comment by situchan** [comment]
I am not able to reproduce. Do you have step?

**Comment by b4s36t4** [comment]
Just click outside of the focus (composer shouldn't be focused) and try paste.

https://github.com/Expensify/App/assets/59088937/168e928f-e48b-4bf9-a5c5-611b96b390e8

**Comment by daordonez11** [comment]
I found a way to reproduce it @situchan 
1. Edit a message
2. Stop editing
3. Click on main composer and paste, you can see the error

In this specific case handler is not being removed right. Currently testing if this is something missing in the migration as well

cc @bondydaa

**Comment by situchan** [comment]
I reproduce it too. This should not happen if using `textInput.current.addEventListener` approach I suggested.

**Comment by b4s36t4** [comment]
@situchan stpes are correct but there is also other issue... now it's pasting the message if compose if not focused?


https://github.com/Expensify/App/assets/59088937/a66a967f-3bdc-4e06-8e14-931430f77802

**Comment by daordonez11** [comment]
Nah its not because of the textInput current is because of the other line I moved, when component is removed textInput.current is undefined so the handler remove is not called.

**Comment by b4s36t4** [comment]
There maybe multiple cases which can cause issues if we un-focus the composer.

**Comment by daordonez11** [comment]
@situchan @bondydaa We can do two things, we can move the remove listener up again or we can target the `textInput.current` element instead of the document.

**Comment by situchan** [comment]
> now it's pasting the message if compose if not focused?

I am still not able to reproduce this

**Comment by b4s36t4** [comment]
But previously it was not... maybe few weeks ago I think

**Comment by b4s36t4** [comment]
What if I copy a file and try to paste without compose focus? I'm pretty use I remember this worked previously.

**Comment by daordonez11** [comment]
@b4s36t4 that was fixed with this issue https://github.com/Expensify/App/issues/22803

**Comment by Santhosh-Sellavel** [comment]
I think a few things got broken [here](https://github.com/Expensify/App/pull/23359). This issue is also regression from the same

**Comment by b4s36t4** [comment]
How about using the `evevnt.target` from the paste event using listener on document level. 

If we have target as `input` or `textarea` we focus that element and handle the paste event.. if not we could directly do the focus on main composer and paste it in that.

**Comment by Pujan92** [comment]
https://github.com/Expensify/App/pull/25419/files#diff-224d9ca075668f353172108020027fb452485426ce8263e0070e71442a642297L373

Order. change seems to be the cause as when we close the edit textinput but `removeEventListener` won't be triggered. So I think we need to revert this specific line.

**Comment by situchan** [comment]
> Order. change seems to be the cause as when we close the edit textinput but removeEventListener won't be triggered. So I think we need to revert this specific line.

yeah that was already discussed. Please check above comments

**Comment by situchan** [comment]
> I think a few things got broken [here](https://github.com/Expensify/App/pull/23359). This issue is also regression from the same

Indeed. It's expected to paste into main composer even when unfocused.
@bondydaa how about raising another follow-up PR here to fix this regression as well?

**Comment by daordonez11** [comment]
This was moved down by me @situchan  just trying to replicate the behavior as it was before, just not sure why now `textInput.current` is undefined. Should we focus on finding why it is undefined or move up the handler again? I created a draft PR in case we want to move it up.

cc @bondydaa

**Comment by bondydaa** [comment]
Let's just link https://github.com/Expensify/App/pull/25424/ to this GH, I don't think we need a new issue since it's just cleaning up something introduced on https://github.com/Expensify/App/pull/25419

**Comment by daordonez11** [comment]
Got it! Follow up PR is ready not sure if it might make sense to leave the comment in the original PR regarding the issue of `textInput.current` being undefined during unmount of the component.

cc @bondydaa @situchan

**Comment by bondydaa** [comment]
we ended up CP'ing https://github.com/Expensify/App/pull/25424 as well and it should be on staging per @daordonez11.

removing the blocker from here so this can go out naturally to production with the next deploy.

**Comment by twisterdotcom** [comment]
Apologies but I am finally going on Parental Leave. I am reassigning to another member of the team.

**Comment by b4s36t4** [comment]
Do I get any reporting bonus https://github.com/Expensify/App/issues/25409#issuecomment-1682729880 for this one?

**Comment by MitchExpensify** [comment]
The issue in question here was reported by our internal testers. Does what @b4s36t4 commented on [here](https://github.com/Expensify/App/issues/25409#issuecomment-1682729880) look like a distinct issue to what was reporting here originally to you @situchan ?

**Comment by situchan** [comment]
> The issue in question here was reported by our internal testers. Does what @b4s36t4 commented on [here](https://github.com/Expensify/App/issues/25409#issuecomment-1682729880) look like a distinct issue to what was reporting here originally to you @situchan ?

I think he's eligible though it's not a formal bug report in slack.

**Comment by MitchExpensify** [comment]
**Payment summary:**

Reporting: Applause N/A
C: @daordonez11 $1500 (Upwork)
C+: @situchan $1500 (Upwork)

**Comment by MitchExpensify** [comment]
[Paid and contracts ended](https://github.com/Expensify/App/issues/25409#issuecomment-1696615082)

**Comment by MitchExpensify** [comment]
Bump on [BZ steps](https://github.com/Expensify/App/issues/25409#issuecomment-1687089997) @situchan

**Comment by situchan** [comment]
- [x] The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/23359
- [x] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/23359#issuecomment-1682529874
- [x] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: N/A
- [x] Determine if we should create a regression test for this bug.
- [x] If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

**Regression Test Proposal:**
1. Open any chat
2. Open any RHP page (i.e. Settings page) and then close
3. Copy any text
4. Paste into the composer
5. Verify that text is not pasted twice

@MitchExpensify I am also eligible for reporting bonus as I reported before this GH is created - https://github.com/Expensify/App/issues/25409#issuecomment-1682464109

**Comment by MitchExpensify** [comment]
> I am also eligible for reporting bonus as I reported before this GH is created - https://github.com/Expensify/App/issues/25409#issuecomment-1682464109

Paid reporting bonus!

**Comment by MitchExpensify** [comment]
Added new test request - Thanks everyone!
